File: src/main/java/day1java/exercise3/EchoMachine.java
package day1java.exercise3;

import java.util.function.Function;
import java.util.function.Supplier;

public class EchoMachine {

    private final Function<String, Console<Unit>> writer;
    private final Supplier<Console<String>> reader;

    public EchoMachine(Function<String, Console<Unit>> writer, Supplier<Console<String>> reader) {
        this.writer = writer;
        this.reader = reader;
    }

    public void echo() {
        Console<String> stringConsole = reader.get();

        String s = stringConsole.exec.get();

        Console<Unit> apply = writer.apply(s);

        apply.exec.get();
    }
}

File: src/main/kotlin/day1/exercise3/EchoMachine.kt
package day1.exercise3

import day1.example3.*

data class EchoMachine(
    val writer: (String) -> Console<Unit>,
    val reader: () -> Console<String>
) {
    fun echo() =
        reader()
            .flatMap { writer(it) }
            .exec()
}

data class Calculator(
    val writer: (String) -> Console<Unit>,
    val reader: () -> Console<String>
) {
    fun run() =
        reader()
            .map {
                parseOperation(it)
                    .map(this::evalOperation)
                    .map { it.toString() }
                    .recoverWith { it.msg }
            }
            .flatMap { writer(it) }
            .exec()

    private fun evalOperation(op: Operation): Int {
        return when (op) {
            is Sum -> op.x + op.y
        }
    }

    private fun parseOperation(input: String): Outcome<ParsingError, Operation> {
        val parts = input.split(' ')
        return if (parts.size != 3)
            Failure(WrongParamsCount("Invalid: $input"))
        else if (parts[0] != "+")
            Failure(UnsupportedOperation("Invalid: $input"))
        else if (parts[1].toIntOrNull() == null ||
            parts[2].toIntOrNull() == null
        )
            Failure(InvalidNumbers("Invalid: $input"))
        else
            Success(Sum(parts[1].toInt(), parts[2].toInt()))
    }


}

sealed class Operation
data class Sum(val x: Int, val y: Int) : Operation()

sealed class ParsingError : Error
data class WrongParamsCount(override val msg: String) : ParsingError()
data class UnsupportedOperation(override val msg: String) : ParsingError()
data class InvalidNumbers(override val msg: String) : ParsingError()
